utilities/flink/flink_compaction_filter.h

```diff
@@ -0,0 +1,159 @@
+
+  class ListElementIter {
+  public:
+    virtual ~ListElementIter() = default;
+    virtual std::size_t NextUnexpiredOffset(const Slice& list, int64_t ttl, int64_t current_timestamp) const = 0;
+  };
+
+  class FixedListElementIter : public ListElementIter {
+  public:
+    explicit FixedListElementIter(std::size_t fixed_size, std::size_t timestamp_offset, std::shared_ptr<Logger> logger) :
+            fixed_size_(fixed_size), timestamp_offset_(timestamp_offset), logger_(std::move(logger)) {}
+    std::size_t NextUnexpiredOffset(const Slice& list, int64_t ttl, int64_t current_timestamp) const override;
+  private:
+      std::size_t fixed_size_;
+      std::size_t timestamp_offset_;
+      std::shared_ptr<Logger> logger_;
+  };
+
+  // Factory is needed to create one iterator per filter/thread
+  // and avoid concurrent access to list state bytes between methods SetListBytes and NextOffset
+  class ListElementIterFactory {
+  public:
+    virtual ~ListElementIterFactory() = default;
+    virtual ListElementIter* CreateListElementIter(std::shared_ptr<Logger> logger) const = 0;
+  };
+
+  class FixedListElementIterFactory : public ListElementIterFactory {
+  public:
+    explicit FixedListElementIterFactory(std::size_t fixed_size, std::size_t timestamp_offset) :
+            fixed_size_(fixed_size), timestamp_offset_(timestamp_offset) {}
+    FixedListElementIter* CreateListElementIter(std::shared_ptr<Logger> logger) const override {
+        return new FixedListElementIter(fixed_size_, timestamp_offset_, logger);
+    };
+  private:
+    std::size_t fixed_size_;
+    std::size_t timestamp_offset_;
+  };
+
+  struct Config {
+    StateType state_type_;
+    std::size_t timestamp_offset_;
+    int64_t ttl_;
+    bool useSystemTime_;
+    ListElementIterFactory* list_element_iter_factory_;
+  };
+
+  // Allows to configure at once all FlinkCompactionFilters created by the factory
+  // which holds the shared ConfigHolder.
+  class ConfigHolder {
+  public:
+    ~ConfigHolder() {
+      Config* config = config_.load();
+      if (config != &DISABLED_CONFIG) {
+        delete config->list_element_iter_factory_;
+        delete config;
+      }
+    }
+
+    // at the moment Flink configures filters (can be already created) only once when user creates state
+    // otherwise it can lead to ListElementIter leak in Config
+    // or race between its delete in Configure() and usage in FilterV2()
+    void Configure(Config* config) {
+      assert(GetConfig() == &DISABLED_CONFIG);
+      config_ = config;
+    }
+
+    void SetCurrentTimestamp(int64_t current_timestamp) {
+      current_timestamp_ = current_timestamp;
+    }
+
+    Config* GetConfig() {
+        return config_.load();
+    }
+
+    std::int64_t GetCurrentTimestamp() {
+      return current_timestamp_.load();
+    }
+
+  private:
+    Config DISABLED_CONFIG = Config{Disabled, 0, std::numeric_limits<int64_t>::max(), true, nullptr};
+    std::atomic<Config*> config_ = { &DISABLED_CONFIG };
```

Comment:
Is this still required to be atomic if we only call configure at a point in the lifecycle before compaction ever runs?

Reply:
I think we still need atomic because when the state is restored in Flink but not configured by user yet, the filter can be already created by Rocksdb and later activated by Flink over atomic while running.